source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -1342,6 +1236,125 @@ bool COpenGLDriver::beginScene(bool backBuffer, bool zBuffer, SColor color,
 }
 
 
+core::smart_refctd_ptr<IGPUShader> COpenGLDriver::createGPUShader(const asset::ICPUShader* _cpushader)
+{
+    return core::make_smart_refctd_ptr<COpenGLShader>(_cpushader);
+}
+
+core::smart_refctd_ptr<IGPUSpecializedShader> COpenGLDriver::createGPUSpecializedShader(const IGPUShader* _unspecialized, const asset::ISpecializationInfo* _specInfo)
+{
+    const COpenGLShader* glUnspec = static_cast<const COpenGLShader*>(_unspecialized);
+    const asset::ICPUShader* cpuUnspec = glUnspec->getCPUCounterpart();
+
+    const std::string& EP = _specInfo->entryPoint;
+    const asset::E_SHADER_STAGE stage = _specInfo->shaderStage;
+
+    const asset::ICPUShader* spvCPUShader = nullptr;
+    if (cpuUnspec->containsGLSL()) {
+        //TODO insert enabled extensions #defines into GLSL
+        auto spvShader = core::smart_refctd_ptr<asset::ICPUShader>(
+            GLSLCompiler->createSPIRVFromGLSL(
+                reinterpret_cast<const char*>(cpuUnspec->getSPVorGLSL()->getPointer()),
+                stage,
+                EP.c_str(),
+                "????"
+            ), core::dont_grab);
+        if (!spvShader)
+            return nullptr;
+
+        spvCPUShader = spvShader.get();
+        spvCPUShader->grab();
+    }
+    else {
+        spvCPUShader = cpuUnspec;
+        spvCPUShader->grab();
+    }
+
+    asset::CShaderIntrospector introspector(GLSLCompiler.get(), { _specInfo->entryPoint, _specInfo->shaderStage });
+    const asset::CIntrospectionData* introspection = introspector.introspect(spvCPUShader);
+    if (introspection->pushConstant.present && introspection->pushConstant.info.name.empty())
+    {
+        assert(false);//abort on debug build
+        os::Printer::log("Attempted to create OpenGL GPU specialized shader from SPIR-V without debug info - unable to set push constants. Creation failed.", ELL_ERROR);
+        return nullptr;//release build: maybe let it return the shader
+    }
+
+    auto ctx = getThreadContext_helper(false);
+    COpenGLSpecializedShader* gpuSpecialized = new COpenGLSpecializedShader(Params.AuxGLContexts + 1u, ctx->ID, this->ShaderLanguageVersion, spvCPUShader->getSPVorGLSL(), _specInfo, introspection);
+    spvCPUShader->drop();
+    return core::smart_refctd_ptr<COpenGLSpecializedShader>(gpuSpecialized, core::dont_grab);
+}
+
+core::smart_refctd_ptr<IGPUBufferView> COpenGLDriver::createGPUBufferView(IGPUBuffer* _underlying, asset::E_FORMAT _fmt, size_t _offset, size_t _size)
+{
+    if (!_underlying)
+        return nullptr;
+    const size_t effectiveSize = (_size != IGPUBufferView::whole_buffer) ? (_underlying->getSize() - _offset) : _size;
+    if ((_offset + effectiveSize) > _underlying->getSize())
+        return nullptr;
+    if ((_offset / reqTBOAlignment)*reqTBOAlignment != _offset) //offset must be aligned to GL_TEXTURE_BUFFER_OFFSET_ALIGNMENT
```

Comment:
you can bank on reqTBOAlignment being a PoT (assert if its not when retrieving it), so you can use one of the fast `isAligned` functions from core::

Reply:
resolved